core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -6553,19 +6595,31 @@ object Exit extends Serializable {
     failCause(Cause.interrupt(id))
 
   def collectAll[E, A](exits: Iterable[Exit[E, A]]): Option[Exit[E, List[A]]] =
-    exits.headOption.map { head =>
-      exits
-        .drop(1)
-        .foldLeft(head.mapExit((a: A) => List(a)))((acc, el) => acc.zipWith(el)((acc, el) => el :: acc, _ ++ _))
-        .mapExit(_.reverse)
-    }
+    collectAllWith(exits)(_ ++ _)
 
   def collectAllPar[E, A](exits: Iterable[Exit[E, A]]): Option[Exit[E, List[A]]] =
-    exits.headOption.map { head =>
-      exits
-        .drop(1)
-        .foldLeft(head.mapExit((a: A) => List(a)))((acc, el) => acc.zipWith(el)((acc, el) => el :: acc, _ && _))
-        .mapExit(_.reverse)
+    collectAllWith(exits)(_ && _)
+
+  private def collectAllWith[E, A](
+    exits: Iterable[Exit[E, A]]
+  )(combineError: (Cause[E], Cause[E]) => Cause[E]): Option[Exit[E, List[A]]] =
+    if (exits.isEmpty) None
+    else {
+      val builder = new ListBuffer[A]
+      var cause   = null.asInstanceOf[Cause[E]]
+      val it      = exits.iterator
+      while (it.hasNext) {
+        val head = it.next()
```

Comment:
Is there a reason to avoid tail recursion  here?

I haven't benchmarked a ListBuffer vs List prepend + reverse

Reply:
ListBuffer is x2 faster and results in half the allocations

```
[info] Benchmark                                           Mode  Cnt      Score     Error   Units
[info] ListBuilderBenchmark.viaLb                          avgt    3    146.439 ±   5.152   ns/op
[info] ListBuilderBenchmark.viaLb:gc.alloc.rate            avgt    3  15628.132 ± 550.647  MB/sec
[info] ListBuilderBenchmark.viaLb:gc.alloc.rate.norm       avgt    3   2400.001 ±   0.001    B/op
[info] ListBuilderBenchmark.viaLb:gc.count                 avgt    3     78.000            counts
[info] ListBuilderBenchmark.viaLb:gc.time                  avgt    3     40.000                ms
[info] ListBuilderBenchmark.viaReverse                     avgt    3    341.447 ±  19.161   ns/op
[info] ListBuilderBenchmark.viaReverse:gc.alloc.rate       avgt    3  13405.135 ± 753.208  MB/sec
[info] ListBuilderBenchmark.viaReverse:gc.alloc.rate.norm  avgt    3   4800.002 ±   0.001    B/op
[info] ListBuilderBenchmark.viaReverse:gc.count            avgt    3     66.000            counts
[info] ListBuilderBenchmark.viaReverse:gc.time             avgt    3     34.000                ms
```

Reply:
Also I know this is a personal preference, but I find while-loops to be more readable than tail recursion with the exception when unpacking lists. But in this case here the input is an `Iterable` so we can't unpack it via tail-recursion